cloud/scope/powervs_cluster.go

```diff
@@ -157,3 +237,1610 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		return nil
+	}
+	for _, lb := range s.IBMPowerVSCluster.Spec.LoadBalancers {
+		if lb.Public {
+			return &lb
+		}
+	}
+	return nil
+}
+
+// SetLoadBalancerStatus set the loadBalancer id.
+func (s *PowerVSClusterScope) SetLoadBalancerStatus(name string, loadBalancer infrav1beta2.VPCLoadBalancerStatus) {
+	if s.IBMPowerVSCluster.Status.LoadBalancers == nil {
+		s.IBMPowerVSCluster.Status.LoadBalancers = make(map[string]infrav1beta2.VPCLoadBalancerStatus)
+	}
+	if val, ok := s.IBMPowerVSCluster.Status.LoadBalancers[name]; ok {
+		if val.ControllerCreated != nil && *val.ControllerCreated {
+			loadBalancer.ControllerCreated = val.ControllerCreated
+		}
+	}
+	s.IBMPowerVSCluster.Status.LoadBalancers[name] = loadBalancer
+}
+
+// GetLoadBalancerID returns the loadBalancer.
+func (s *PowerVSClusterScope) GetLoadBalancerID(loadBalancer infrav1beta2.VPCLoadBalancerSpec) *string {
+	if loadBalancer.Name == "" || s.IBMPowerVSCluster.Status.LoadBalancers == nil {
+		return nil
+	}
+	if val, ok := s.IBMPowerVSCluster.Status.LoadBalancers[loadBalancer.Name]; ok {
+		return val.ID
+	}
+	return nil
+}
+
+// GetLoadBalancerState will get the state for the load balancer.
+func (s *PowerVSClusterScope) GetLoadBalancerState(name string) *infrav1beta2.VPCLoadBalancerState {
+	if s.IBMPowerVSCluster.Status.LoadBalancers == nil {
+		return nil
+	}
+	if val, ok := s.IBMPowerVSCluster.Status.LoadBalancers[name]; ok {
+		return &val.State
+	}
+	return nil
+}
+
+// GetLoadBalancerHost will return the hostname of load balancer.
+func (s *PowerVSClusterScope) GetLoadBalancerHost(name string) *string {
+	if s.IBMPowerVSCluster.Status.LoadBalancers == nil {
+		return nil
+	}
+	if val, ok := s.IBMPowerVSCluster.Status.LoadBalancers[name]; ok {
+		return val.Hostname
+	}
+	return nil
+}
+
+// SetCOSBucketStatus set the COS bucket id.
+func (s *PowerVSClusterScope) SetCOSBucketStatus(cosInstanceID string, controllerCreated bool) {
+	if s.IBMPowerVSCluster.Status.COSInstance == nil {
+		s.IBMPowerVSCluster.Status.COSInstance = &infrav1beta2.ResourceReference{
+			ID:                &cosInstanceID,
+			ControllerCreated: &controllerCreated,
+		}
+	} else {
+		// set the controllerCreated only if its not set
+		if !*s.IBMPowerVSCluster.Status.COSInstance.ControllerCreated {
+			s.IBMPowerVSCluster.Status.COSInstance.ControllerCreated = &controllerCreated
+		}
+		s.IBMPowerVSCluster.Status.COSInstance.ID = &cosInstanceID
+	}
+}
+
+// GetResourceGroupID returns the resource group id.
+func (s *PowerVSClusterScope) GetResourceGroupID() (string, error) {
+	if s.IBMPowerVSCluster.Spec.ResourceGroup != nil && s.IBMPowerVSCluster.Spec.ResourceGroup.ID != nil {
+		return *s.IBMPowerVSCluster.Spec.ResourceGroup.ID, nil
+	}
+	if s.IBMPowerVSCluster.Status.ResourceGroup != nil && s.IBMPowerVSCluster.Status.ResourceGroup.ID != nil {
+		return *s.IBMPowerVSCluster.Status.ResourceGroup.ID, nil
+	}
+	resourceGroupID, err := s.fetchResourceGroupID()
```

Comment:
maybe handle this in a better way by avoiding calling `fetchResourceGroupID` here